### ISSUE TITLE
Local Runtime (Java) Memory Leak

### ISSUE THREAD

**Issue by No3371** [spatialos/gdk-for-unity#1400]
#### Description
I am working on managed AI worker which spawns 715 entities, modify the position components and let SpatialOS update the data to runtime, every frame.

If I disable the worker (by disabling the worker connector before building it), and run the worker in Unity Editor, the memory usage stay stable at about 4GB memory working fine.

If I enable the worker connector and let it managed by the local runtime, the Java application consumes all my memory.

I've isolated the issue by testing the manged worker with/without the system that assign new value to Position.

#### Expected behaviour
The worker works normally whether it's managed/external.
(I expected workers running inside Unity should have more overhead)

#### Environment
i7-9700 / 32GB
Windows 10 1909 / Unity 2019.3.7f1
OpenJDK Runtime Environment (build 14.0.1+7)
OpenJDK 64-Bit Server VM (build 14.0.1+7, mixed mode, sharing)

**Comment by No3371** [comment]
> I am working on managed AI worker which spawns 715 entities, modify the position components and let SpatialOS update the data to runtime, **every frame**.

So it turns out it's because a managed worker runs way faster then one running inside UnityEditor, so it generate enormous traffic toward the runtime.

After limit `targetFrameRate` to 60, the memory usage stays at about 4GB.

**Comment by No3371** [comment]
Unfortunately I have to reopen this since the memory usage will randomly grows, like, after staying at 4GB for 5 minutes, it suddenly goes up to 13GB, then after 2 minutes, it goes up to 15GB.

**Comment by No3371** [comment]
> [DONE] Managed worker opening inspector (Issue reproduce in 8 mins, memory usage grows to 9GB)
I've reproduced the random memory leak, it happens when Inspector opened.

Another minor issue along with this:
Under heavy load Entity count in Inspector is being decreased overtime. The entity count decreased to 6XX, but shows 715 if I opens with Chrome at the time.

**Comment by zeroZshadow** [comment]
Does this happen with just the Inspector, or also Inspector V2?
When you mention CPU and memory usage, is this for the worker or the runtime process?

Note that sening an update on `Position` is special and rather heavy, as this will cause the runtime to re-evaluate all subscriptions made through QBI queries.

**Comment by No3371** [comment]
> Does this happen with just the Inspector, or also Inspector V2?
I did not tested with InspectorV2, I've tried it multiple times though, but it's not as intuitive as V1 😅 

> When you mention CPU and memory usage, is this for the worker or the runtime process?
It's the java process. It always occupy a lot of CPU usage, I am thinking it may be related to what you said below.

> Note that sening an update on `Position` is special and rather heavy, as this will cause the runtime to re-evaluate all subscriptions made through QBI queries.
Thank you, that's very valuable info!
Does this means I can give up Immediate updating of Inspector positions and use custom position component for performance?

**Comment by zeroZshadow** [comment]
Yes it is recommended to only update `Position` at a low rate, and use a custom component for the actual position of the entity.
`Position` is used for things like queries, the inspector, and for the initial prefab spawn if you use the default gameobject creator.

**Comment by No3371** [comment]
OK, thanks for the heads up.

I am in the progress to finish up a demo project and moving to cloud, so may not be able to reproduce the issue in near future, as we yet have a conclusion, I'll let this remain opened, will definitely report back if it happens again.
